src/hotspot/share/opto/intrinsicnode.cpp

```diff
@@ -261,27 +261,41 @@ static const Type* bitshuffle_value(const TypeInteger* src_type, const TypeInteg
   }
 
   if (!mask_type->is_con()) {
-    int mask_max_bw;
-    int max_bw = bt == T_INT ? 32 : 64;
-    // Case 1) Mask value range includes -1.
-    if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
-      mask_max_bw = max_bw;
-    // Case 2) Mask value range is less than -1.
-    } else if (mask_type->hi_as_long() < -1L) {
-      mask_max_bw = max_bw - 1;
-    } else {
-    // Case 3) Mask value range only includes +ve values.
-      assert(mask_type->lo_as_long() >= 0, "");
-      jlong clz = count_leading_zeros(mask_type->hi_as_long());
-      clz = bt == T_INT ? clz - 32 : clz;
-      mask_max_bw = max_bw - clz;
-    }
     if ( opc == Op_CompressBits) {
-      lo = mask_max_bw == max_bw ? lo : 0L;
-      // Compress operation is inherently an unsigned operation and
-      // result value range is primarily dependent on true count
-      // of participating mask value.
-      hi = mask_max_bw < max_bw ? (1L << mask_max_bw) - 1 : src_type->hi_as_long();
+      // Pattern: Integer/Long.compress(src_type, mask_type)
+      int max_mask_bit_width;
+      int mask_bit_width = bt == T_INT ? 32 : 64;
+      if ((mask_type->lo_as_long() < 0L && mask_type->hi_as_long() >= -1L)) {
+        // Case 1) Mask value range includes -1, this negates the possibility of
+        // strictly non-negative result value range.
+        max_mask_bit_width = mask_bit_width;
+      } else if (mask_type->hi_as_long() < -1L) {
+        // Case 2) Mask value range is less than -1, this indicates presence of at least
+        // one zero bit in the mask value, there by constraining the result of compression
+        // to a +ve value range.
+        max_mask_bit_width = mask_bit_width - 1;
+      } else {
+        // Case 3) Mask value range only includes +ve values, this can again be
+        // used to ascertain known Zero bits of resultant value.
+        assert(mask_type->lo_as_long() >= 0, "");
+        // Here, result of clz is w.r.t to long argument, hence for integer argument
+        // we explicitly subtract 32 from the result.
+        jlong clz = count_leading_zeros(mask_type->hi_as_long());
+        clz = bt == T_INT ? clz - 32 : clz;
+        max_mask_bit_width = mask_bit_width - clz;
+      }
+
+      // If number of bits required to accommodated mask value range is less than the
+      // full bit width of integral type, then MSB bit is guaranteed to be zero, thus
+      // compression result will never be a -ve value and we can safely set the
+      // lower bound of the result value range to zero.
+      lo = max_mask_bit_width == mask_bit_width ? lo : 0L;
```

Comment:
Can you please add an assert that we are not making `lo` worse than what we already have? Someone may insert optimizations above that set `lo > 0`, and then you may lower it again here.
```suggestion
      assert(lo < 0, "we should not lower the value of lo");
      lo = max_mask_bit_width == mask_bit_width ? lo : 0L;
```

Reply:
We are already initializing 'lo' to min_jint/jlong value upfront and nothing before this line is modifying its value. Since initialization and use both are part of this function hense adding an assertion looks redundant. We generally add assertions to set constraints under which a logic is implemented.

Reply:
Your response is saying that the code is currently correct. That is true as far as I can see now.
But my worry is about future changes. And an assert can also help the reader be sure that things happen correctly without having to verify manually, which is a lot of effort. So for me it would not be redundant.